scapy/layers/ppp.py

```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+
+_PPP_paptypes = {1: "Authenticate-Request",
+                 2: "Authenticate-Ack",
+                 3: "Authenticate-Nak"}
+
+
+class PPP_PAP(Packet):
+    name = "PPP Password Authentication Protocol"
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id",0),
+                   ShortField("len", None),
+                   StrLenField("data","",length_from=lambda p:p.len-4)]
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *_, **kargs):
+        code = None
+        if _pkt:
+            code = ord(_pkt[0])
+        elif "code" in kargs:
+            code = kargs["code"]
+            if isinstance(code, basestring):
+                code = cls.fields_desc[0].s2i[code]
+
+        if code == 1:
+            return PPP_PAP_Request
+        elif code in [2,3]:
+            return PPP_PAP_Response
+        return cls
+
+    def extract_padding(self, pay):
+        return "",pay
+
+    def post_build(self, p, pay):
+        p += pay
+        if self.len is None:
+            p = p[:2]+struct.pack("!H", len(p))+p[4:]
+        return p
+
+
+class PPP_PAP_Request(PPP_PAP):
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   FieldLenField("username_len", None, fmt="B", length_of="username"),
+                   StrLenField("username", None, length_from=lambda p: p.username_len),
+                   FieldLenField("passwd_len", None, fmt="B", length_of="password"),
+                   StrLenField("password", None, length_from=lambda p: p.passwd_len)]
+
+    def mysummary(self):
+        return self.sprintf("PAP-Request username=%PPP_PAP_Request.username% password=%PPP_PAP_Request.password%")
+
+
+class PPP_PAP_Response(PPP_PAP):
+    fields_desc = [ByteEnumField("code", 2, _PPP_paptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   FieldLenField("msg_len", None, fmt="B", length_of="message"),
+                   StrLenField("message", "", length_from=lambda p: p.msg_len)]
+
+    def answers(self, other):
+        return isinstance(other, PPP_PAP_Request) and other.id == self.id
+
+    def mysummary(self):
+        res = "PAP-Ack" if self.code == 2 else "PAP-Nak"
+        if self.msg_len > 0:
+            res += self.sprintf(" msg=%PPP_PAP_Response.message%")
+        return res
+
+
+### Challenge Handshake Authentication protocol (RFC1994)
+
+_PPP_chaptypes = {1: "Challenge",
+                  2: "Response",
+                  3: "Success",
+                  4: "Failure"}
+
+
+class PPP_CHAP(Packet):
+    name="PPP Challenge Handshake Authentication Protocol"
+    fields_desc = [ByteEnumField("code", 1, _PPP_chaptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   ConditionalField(StrLenField("data","",length_from=lambda p: p.len-4),
```

Comment:
Since both fields are `StrLenField`, I think it would be better to remove these `ConditionalField` and use the same name in both cases. What do you think?
Also, this would ease the use of a `FieldLenField` for `"len"`.